## Session #1 Homework

In [1]:
import numpy as np
import pandas as pd

### Question 1

In [2]:
np.__version__

'1.21.5'

In [3]:
df = pd.read_csv("./data.csv")

**How many records are in the dataset?**

In [4]:
df.shape

(11914, 16)

There are 11914 rows and 16 columns

**Who are the most popular car manufactures (top-3) according to the dataset?**

In [5]:
df.groupby("Make").sum().Popularity.nlargest(3)

Make
Ford         4983817
Chevrolet    1555355
Toyota       1515126
Name: Popularity, dtype: int64

**What's the number of unique Audi car models in the dataset?**

In [6]:
df[df["Make"] == "Audi"]["Model"].nunique()

34

**How many columns in the dataset have missing values?**

In [7]:
count = 0
for col in df.columns:
    if df[col].isnull().sum() >0 :
        count+=1
print(f"No of missing columns = {count}")

No of missing columns = 5


**Question 6**
* Find the median value of "Engine Cylinders" column in the dataset 

In [8]:
df["Engine Cylinders"].median()

6.0

* Calculate the most frequent value of the same "Engine Cylinders"

In [9]:
mode=df["Engine Cylinders"].mode()
mode

0    4.0
Name: Engine Cylinders, dtype: float64

* Use the `fillna` method to fill the missing valuues in "Engine Cylinders" with the most frequent value from the previous step

In [10]:
df["Engine Cylinders"].fillna(value=4,inplace = True)

In [11]:
df["Engine Cylinders"].isnull().sum()

0

* Calculate the median value of "Engine Cylinders" once again

In [12]:
df["Engine Cylinders"].median()

6.0

**Question 7**

* Select all "Lotus" cars from the dataset.

In [23]:
sub_df = df[df["Make"] == "Lotus"]

* Select only columns "Engine HP", "Engine Cylinders"

In [24]:
sub_df = sub_df[["Engine HP","Engine Cylinders"]]

In [25]:
sub_df.columns

Index(['Engine HP', 'Engine Cylinders'], dtype='object')

* Drop all duplicated rows using `drop_duplicates` method

In [26]:
sub_df = sub_df.drop_duplicates()

* Ge the underlying NumPy array

In [27]:
X = sub_df.to_numpy()
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

* Compute matrix-matrix multiplication between the transpose of X and X

In [28]:
XTX = np.matmul(X.T,X)

* Invert XTX

In [29]:
inv_XTX = np.linalg.inv(XTX)

* Create an array

In [30]:
y = np.array([1100,800,750,850,1300,1000,1000,1300,800])

* Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`

In [31]:
i = np.matmul(inv_XTX,X.T)
w = np.matmul(i,y)

* What's the value of the first element of `w`?

In [32]:
w[0]

4.594944810094579